test/tests/parallel/set1/spam_virus/fail2ban.bats

```diff
@@ -0,0 +1,203 @@
+  common_container_setup 'CUSTOM_SETUP_ARGUMENTS'
+}
+
+function teardown_file() {
+  docker rm -f "${CONTAINER1_NAME}" "${CONTAINER2_NAME}"
+}
+
+@test "${TEST_NAME_PREFIX} Fail2Ban is running" {
+  _run_in_container /bin/bash -c "ps aux --forest | grep -v grep | grep '/usr/bin/python3 /usr/bin/fail2ban-server'"
+  assert_success
+}
+
+@test "${TEST_NAME_PREFIX} localhost is not banned because ignored" {
+  _run_in_container /bin/sh -c "fail2ban-client status postfix-sasl | grep 'IP list:.*127.0.0.1'"
+  assert_failure
+  _run_in_container grep 'ignoreip = 127.0.0.1/8' /etc/fail2ban/jail.conf
+  assert_success
+}
+
+@test "${TEST_NAME_PREFIX} fail2ban-fail2ban.cf overrides" {
+  _run_in_container fail2ban-client get loglevel
+  assert_success
+  assert_output --partial 'DEBUG'
+}
+
+@test "${TEST_NAME_PREFIX} fail2ban-jail.cf overrides" {
+  for FILTER in 'dovecot' 'postfix' 'postfix-sasl'
+  do
+    _run_in_container /bin/sh -c "fail2ban-client get ${FILTER} bantime"
+    assert_output 1234
+
+    _run_in_container /bin/sh -c "fail2ban-client get ${FILTER} findtime"
+    assert_output 321
+
+    _run_in_container /bin/sh -c "fail2ban-client get ${FILTER} maxretry"
+    assert_output 2
+
+    _run_in_container /bin/sh -c "fail2ban-client -d | grep -F \"['set', 'dovecot', 'addaction', 'nftables-multiport']\""
+    assert_output "['set', 'dovecot', 'addaction', 'nftables-multiport']"
+
+    _run_in_container /bin/sh -c "fail2ban-client -d | grep -F \"['set', 'postfix', 'addaction', 'nftables-multiport']\""
+    assert_output "['set', 'postfix', 'addaction', 'nftables-multiport']"
+
+    _run_in_container /bin/sh -c "fail2ban-client -d | grep -F \"['set', 'postfix-sasl', 'addaction', 'nftables-multiport']\""
+    assert_output "['set', 'postfix-sasl', 'addaction', 'nftables-multiport']"
+  done
+}
+
+@test "${TEST_NAME_PREFIX} ban ip on multiple failed login" {
+  # can't pipe the file as usual due to postscreen
+  # respecting postscreen_greet_wait time and talking in turn):
+
+  # shellcheck disable=SC1004
+  for _ in {1,2}
+  do
+    docker exec "${CONTAINER2_NAME}" /bin/bash -c \
+    'exec 3<>/dev/tcp/${MAIL_FAIL2BAN_IP}/25 && \
+    while IFS= read -r cmd; do \
+      head -1 <&3; \
+      [[ ${cmd} == "EHLO"* ]] && sleep 6; \
+      echo ${cmd} >&3; \
+    done < "/tmp/docker-mailserver-test/auth/smtp-auth-login-wrong.txt"'
+  done
+
+  # need a bit more sleep as the test below that ´
+  # restarts F2B might interfere otherwise
+  sleep 30
+
+  # Checking that CONTAINER2_IP is banned in "${CONTAINER1_NAME}"
+  _run_in_container fail2ban-client status postfix-sasl
+  assert_success
+  assert_output --partial "${CONTAINER2_IP}"
+
+  # Checking that CONTAINER2_IP is banned by nftables
+  _run_in_container nft list set inet f2b-table addr-set-postfix-sasl
+  # BATS can reorder tests when running in parallel, which means we cannot use
+  # 'elements = { ${CONTAINER2_IP} }' because there might be other IPs already
+  # banned.
```

Comment:
Are the tests cases within the same file running at the same time?

My recollection was that multiple `.bats` files were being run in parallel on a system, not individual test cases.

Reply:
Haha, see my answer above: <https://github.com/docker-mailserver/docker-mailserver/pull/2938#discussion_r1056239775> :D